### PR TITLE
Reworked and optimized reconnect logic

### PR DESCRIPTION
**Pull request by htbmw**
Reworked and optimized reconnect logic to remove redundancy and fix reconnect issue that occurred since 1.4.368.33 release of OPCFoundation packages

### CODE REVIEW

File: src/Infrastructure/OpcUa/OpcSession.cs [hunk 1]
```diff
@@ -35,14 +34,13 @@ namespace OMP.Connector.Infrastructure.OpcUa
...
         private readonly OpcUaConfiguration _opcUaSettings;
-        private readonly IOpcSessionReconnectHandlerFactory _opcSessionReconnectHandlerFactory;
         private readonly ILoggerFactory _loggerFactory;
         private readonly ILogger _logger;
         private readonly ApplicationConfiguration _applicationConfiguration;
         private bool _disposed;
         private SemaphoreSlim _opcSessionSemaphore;
         private CancellationTokenSource _sessionCancellationTokenSource;
-        private IOpcSessionReconnectHandler _reconnectHandler;
+        private SessionReconnectHandler _sessionReconnectHandler;
```
Comment (ivomp) [inline src/Infrastructure/OpcUa/OpcSession.cs:43]: Why move away from the interface? This introduces hard coupling again with no way to easily extend it.

Comment (ivomp) [inline src/Infrastructure/OpcUa/OpcSession.cs:43]: Why move away from the interface? This introduces hard coupling again with no way to easily extend it.

File: src/Infrastructure/OpcUa/OpcSession.cs [hunk 2]
```diff
@@ -128,6 +125,37 @@ private async Task ConnectAsync(EndpointDescription endpointDescription, IUserId
...
+            if (e.Status != null && ServiceResult.IsNotGood(e.Status))
+            {
+                _logger.Warning($"Communication error: [{e?.Status}] on Endpoint: [{sender.Endpoint?.EndpointUrl}]");
+
+                if (_opcUaSettings.ReconnectIntervalInSeconds <= 0)
+                    return;
+
+                if (_sessionReconnectHandler == null)
+                {
+                    _sessionReconnectHandler = new SessionReconnectHandler();
```
Comment (ivomp) [inline src/Infrastructure/OpcUa/OpcSession.cs:139]: Tight coupling. Can we not then add this logic directly into the OpcSession object?

Comment (ivomp) [inline src/Infrastructure/OpcUa/OpcSession.cs:139]: Tight coupling. Can we not then add this logic directly into the OpcSession object?

File: src/Infrastructure/OpcUa/OpcSession.cs (hunk 1 above)
Comment (htbmw) [reply src/Infrastructure/OpcUa/OpcSession.cs:43]: The interface and implementation was an extra layer that anyway wrapped the SessionReconnectHandler.
The restore of subscriptions is now natively handled by the stack and there is no need for the extra layer. The layer / wrapper is now redundant.

File: src/Infrastructure/OpcUa/OpcSession.cs (hunk 2 above)
Comment (htbmw) [reply src/Infrastructure/OpcUa/OpcSession.cs:139]: Not sure what you mean. Are you referring to dependency injection?

Comment (ivomp) [reply src/Infrastructure/OpcUa/OpcSession.cs:139]: ` _sessionReconnectHandler = new SessionReconnectHandler();`

Marries these 2 objects. Why did the interface and factory break?

File: src/Infrastructure/OpcUa/OpcSession.cs (hunk 1 above)
Comment (htbmw) [reply src/Infrastructure/OpcUa/OpcSession.cs]: Restored the interface and wrapper

Comment (htbmw) [reply src/Infrastructure/OpcUa/OpcSession.cs]: Restored the interface and wrapper

File: src/Infrastructure/OpcUa/OpcSession.cs (hunk 2 above)
Comment (htbmw) [reply src/Infrastructure/OpcUa/OpcSession.cs]: Restored the interface and wrapper

Comment (htbmw) [reply src/Infrastructure/OpcUa/OpcSession.cs]: Restored the interface and wrapper

**Review by ivomp — APPROVED** [review]

**Pull request merged**